server/server/src/file_service.rs

```diff
@@ -1,213 +1,242 @@
+use redis_utils::{tx, TxError};
+use uuid::Uuid;
 
 pub async fn upsert_file_metadata(
     context: RequestContext<'_, FileMetadataUpsertsRequest>,
 ) -> Result<(), ServerError<FileMetadataUpsertsError>> {
     let (request, server_state) = (&context.request, context.server_state);
-    let mut transaction = match server_state.index_db_client.begin().await {
-        Ok(t) => t,
-        Err(e) => {
-            return Err(InternalError(format!("Cannot begin transaction: {:?}", e)));
+    check_for_changed_root(&request.updates)?;
+
+    let mut con = server_state.index_db_pool.get().await?;
+    let mut docs_to_delete: Vec<EncryptedFileMetadata> = vec![];
+    // TODO should we further check that each metadata has the right owner?
+    let tx = tx!(&mut con, pipe, &[owned_files(&context.public_key)], {
+        let now = get_time().0 as u64;
+        let files: Vec<Uuid> = con
+            .maybe_json_get(owned_files(&context.public_key))
+            .await?
+            .ok_or(Abort(ClientError(FileMetadataUpsertsError::UserNotFound)))?;
+        let keys: Vec<String> = files.into_iter().map(keys::file).collect();
+        let mut files: Vec<EncryptedFileMetadata> = con.watch_json_mget(keys).await?;
+
+        docs_to_delete = apply_changes(now, &request.updates, &mut files)?;
+
+        files
+            .verify_integrity()
+            .map_err(|_| Abort(ClientError(GetUpdatesRequired)))?;
+
+        for the_file in &files {
+            pipe.json_set(file(the_file.id), the_file)?;
+            if the_file.deleted && the_file.file_type == Document {
+                pipe.del(size(the_file.id));
+            }
         }
-    };
+        pipe.json_set(owned_files(&context.public_key), files.ids())?;
+        Ok(&mut pipe)
+    });
+    return_if_error!(tx);
 
-    for upsert in &request.updates {
-        if let Some((old_parent, _)) = upsert.old_parent_and_name {
-            // prevent all updates to root
-            if upsert.id == old_parent {
-                return Err(ClientError(FileMetadataUpsertsError::GetUpdatesRequired));
-                // todo: better error
+    for file in docs_to_delete {
+        file_content_client::delete(&server_state.files_db_client, file.id, file.content_version)
+            .await
+            .map_err(|err| internal!("Cannot delete file in S3: {:?}", err))?;
+    }
+    Ok(())
+}
+
+fn check_for_changed_root(
+    changes: &[FileMetadataDiff],
+) -> Result<(), ServerError<FileMetadataUpsertsError>> {
+    for change in changes {
+        if let Some((old_parent, _)) = change.old_parent_and_name {
+            if change.id == old_parent {
+                return Err(ClientError(RootImmutable));
             }
-            // prevent turning existing folder into root
-            if upsert.id == upsert.new_parent {
-                return Err(ClientError(FileMetadataUpsertsError::GetUpdatesRequired));
-                // todo: better error
+            if change.id == change.new_parent {
```

Comment:
unclear how this would be resolved by getting updates; seems like this is a case of a folder being moved into itself which shouldn't be allowed even locally

Reply:
I believe this is the same behavior as prior functionality, I think all the get_updates can have better variants for debugability. There are lots of instances of `GetUpdatesRequired` being returned for things that no updating would solve (for example root changes)